### PR TITLE
Run unit tests on GitHub Actions.

### PR DESCRIPTION
**Pull request by hvy**
## Motivation

Partially takes over #1343. CI on CircleCI is time consuming and we might be able to reduce this time by running jobs concurrently. Instead of replacing existing CI, we can try out a concurrent version on GitHub Actions. Separating the integration module tests and running them simultaneously I could reduce the time to about 9m for the longest workflow (https://github.com/hvy/optuna/actions/runs/130810943, https://github.com/hvy/optuna/actions/runs/130810921) from roughly 20-30m. See also, comments below.

## Description of the changes

Adds GitHub Actions to run existing units tests. Integration module tests are separated into a separate workflow so that workflows can be executed concurrently. CircleCI is left untouched so that we can try both for a period.

#### Note

- It's quite straight forward to condition the integration module workflow to only run when certain integration files are modified using https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#onpushpull_requestpaths if we want to. 
- GitHub Action workflows cannot be canceled automatically on e.g. new pushes.

### CODE REVIEW

**Comment by hvy** [comment]
Thanks for going through this PR. 

Yes the `setup.py` changes are somewhat ad-hoc. I was thinking about following this PR up another one cleaning up the different extras sections. Do you suggest we leave out the changes in `setup.py` out from this PR? It's quite important to cut down the installation time.

**Comment by crcrpar** [comment]
> Do you suggest we leave out the changes in `setup.py` out from this PR?

No, I don't. I'm concerned that some libraries are referenced multiple times in `setup.py` and that makes `setup.py` look confusing when editing. Therefore I feel the need of refactoring.

**Comment by hvy** [comment]
So we are on the same page then. Sorry for the somewhat confusing question. Let me know if changes are needed to this PR. From my side, it's "PTAL".

File: setup.py [hunk 1]
```diff
@@ -122,6 +122,40 @@ def get_extras_require() -> Dict[str, List[str]]:
...
+            "torchvision==0.5.0" if sys.platform == "darwin" else "torchvision==0.5.0+cpu",
+            "xgboost",
+        ]
+        + (
+            ["allennlp<1", "fastai<2", "pytorch-lightning>=0.7.1"]
+            if (3, 5) < sys.version_info[:2] < (3, 8)
+            else []
+        )
+        + (
+            ["keras", "tensorflow", "tensorflow-datasets"] if sys.version_info[:2] < (3, 8) else []
```
Comment (crcrpar) [inline setup.py:157]: Don't we need to avoid Keras 2.4 as https://github.com/optuna/optuna/pull/1386?

Comment (hvy) [reply setup.py]: You're absolutely right. Fixed as suggested. Also merged with master.

**Comment by hvy** [comment]
PTAL (c.f. https://github.com/optuna/optuna/pull/1352#discussion_r445305236).

File: setup.py [hunk 2]
```diff
@@ -132,6 +132,42 @@ def get_extras_require() -> Dict[str, List[str]]:
...
+            "cma",
+            "lightgbm",
+            "mlflow",
+            "mpi4py",
+            "mxnet",
+            "pandas",
+            "pytorch-ignite",
+            "scikit-learn>=0.19.0,<0.23.0",
+            "scikit-optimize",
+            "torch==1.4.0" if sys.platform == "darwin" else "torch==1.4.0+cpu",
```
[the remark this answers is not in the document]
Comment (hvy) [reply setup.py]: Fixed.

File: setup.py [hunk 3]
```diff
@@ -132,6 +132,42 @@ def get_extras_require() -> Dict[str, List[str]]:
...
+            "torchvision==0.5.0" if sys.platform == "darwin" else "torchvision==0.5.0+cpu",
```
[the remark this answers is not in the document]
Comment (hvy) [reply setup.py]: Fixed.

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    - name: Setup Python${{ matrix.python-version }}
+      uses: actions/setup-python@v1
+      with:
+        python-version: ${{ matrix.python-version }}
+
+    - name: Setup environment
+      run: |
+        sudo apt-get update
+
+        # Install swig which is required by pyrfr which is required by fanova.
```
[the remark this answers is not in the document]
Comment (hvy) [reply .github/workflows/tests.yml]: Thanks, fixed.

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        pip install --progress-bar off .
+        python -c 'import optuna'
+        optuna --version
+
+        pip install --progress-bar off .[tests]
+        pip install --progress-bar off .[optional]
+
+        # Required for samplers_tests/test_samplers.py.
+        # TODO(hvy): Fix samplers_tests/test_samplers.py to not require optional depenendencies and remove these installs.
+        pip install scikit-optimize
```
[the remark this answers is not in the document]
Comment (hvy) [reply .github/workflows/tests.yml:54]: Hm, yes I think so. `skopt` is still required by this tests.

**Comment by hvy** [comment]
Thanks for reviewing. Some changes were obsolete as you pointed out. Comments should be addressed now. PTAL.

File: .github/workflows/tests.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (himkt) [reply .github/workflows/tests.yml:54]: My apologies, I confused `tests.ym` and `tests-integration.yml`.

**Comment by hvy** [comment]
I'd like to follow up this PR with an additional daily cron scheduled triggered. That way, we can detect failures caused by upgraded dependencies on a regular basis.

File: .github/workflows/tests-integration.yml [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  build:
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        python-version: [3.5, 3.6, 3.7, 3.8]
+
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
```
Comment (toshihikoyanase) [inline .github/workflows/tests-integration.yml:19]: [Question] Do you expect to execute the CI jobs in the forked repositories?
If not we can stop it as follows:
```suggestion
    # This action cannot be executed in the forked repository.
    if: github.repository == 'optuna/optuna'
    steps:
    - name: Checkout
      uses: actions/checkout@v2
```

Comment (hvy) [reply .github/workflows/tests-integration.yml:22]: Thanks, fixed in https://github.com/optuna/optuna/pull/1352/commits/d18aa2a3d1a03c30d2f3a700dab084bb5f14ea8d.

**Comment by hvy** [comment]
PTAL.

**Pull request merged**